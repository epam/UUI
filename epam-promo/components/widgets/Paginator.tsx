import { Paginator as UuiPaginator, PaginatorParams, PaginatorProps } from '@epam/uui-components';
import React from 'react';
import { PageButton } from './PageButton';
import css from './Paginator.scss';
import { ReactComponent as ArrowLeftIcon_24 } from '@epam/assets/icons/common/navigation-chevron-left-18.svg';
import { ReactComponent as ArrowRightIcon_24 } from '@epam/assets/icons/common/navigation-chevron-right-18.svg';
import { ReactComponent as ArrowLeftIcon_30 } from '@epam/assets/icons/common/navigation-chevron-left-24.svg';
import { ReactComponent as ArrowRightIcon_30 } from '@epam/assets/icons/common/navigation-chevron-right-24.svg';

export class Paginator extends React.Component<PaginatorProps> {
    renderPaginator(params: PaginatorParams) {
        return (
            <nav role="navigation" className={css.root} {...params.rawProps}>
                <PageButton
                    size={params.size}
                    icon={params.size === '30' ? ArrowLeftIcon_30 : ArrowLeftIcon_24}
                    onClick={params.goToPrev}
                    isDisabled={params.isFirst}
                    fill="white"
                    color="gray50"
                />
                {params.pages.map((page, index) => {
                    if (page.type === 'spacer') {
                        return <PageButton size={params.size} key={`${index}_spacer`} caption={'...'} fill="light" color="blue" tabIndex={-1} />;
                    } else {
                        return (
                            <PageButton
                                size={params.size}
                                key={page.pageNumber}
                                caption={page.pageNumber}
                                onClick={() => page.onClick?.()}
                                rawProps={{ 'aria-current': page.isActive }}
                                fill={(page.isActive && 'white') || 'light'}
                                color={'blue'}
                            />
                        );
                    }
                })}
                <PageButton
                    size={params.size}
                    icon={params.size === '30' ? ArrowRightIcon_30 : ArrowRightIcon_24}
                    onClick={params.goToNext}
                    isDisabled={params.isLast}
                    fill="white"
                    color="gray50"
                />
            </nav>
        );
    }

    render() {
        return <UuiPaginator {...this.props} render={this.renderPaginator} />;
    }
}
