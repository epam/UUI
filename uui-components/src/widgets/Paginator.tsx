import * as React from 'react';
import range from 'lodash.range';
import { IAnalyticableOnChange, IEditable, UuiContext, UuiContexts } from "@epam/uui";

export interface PaginatorParams {
    size: '24' | '30';
    pages: PaginatorItem[];
    goToNext(): void;
    goToPrev(): void;
    isLast: boolean;
    isFirst: boolean;
}

interface PaginatorItem {
    type: 'page' | 'spacer';
    pageNumber?: number;
    isActive?: boolean;
    onClick?(): void;
}

export interface PaginatorProps extends IEditable<number>, IAnalyticableOnChange<number> {
    size: '24' | '30';
    totalPages: number;
    render?(params: PaginatorParams): any;
}

const FIRST_PAGE = 1;
const SIMPLE_PAGINATION_ITEMS = 7;

export class Paginator extends React.Component<PaginatorProps> {
    static contextType = UuiContext;
    context: UuiContexts;
    // size = this.props.size || '36';

    isFirst = () => {
        return this.props.value === FIRST_PAGE;
    }

    isLast = () => {
        return this.props.value === this.props.totalPages;
    }

    getPagesView(): PaginatorItem[] {
        const paginatorItems: PaginatorItem[] = [];
        const pages = this.props.totalPages;
        const currentPage = this.props.value;
        const onClick = (value: number) => {
            this.props.onValueChange(value);

            if (this.props.getValueChangeAnalyticsEvent) {
                const event = this.props.getValueChangeAnalyticsEvent(value, this.props.value);
                this.context.uuiAnalytics.sendEvent(event);
            }
        };

        function addPage(page: number) {
            if (page !== currentPage) {
                return paginatorItems.push({ type: 'page', pageNumber: page, onClick: () => onClick(page) });
            }
            return paginatorItems.push({ type: 'page', pageNumber: page, onClick: () => onClick(page), isActive: true });
        }

        function addSpacer() {
            return paginatorItems.push({ type: 'spacer' });
        }

        // If the number of pages is not more than the maximum number of displayed pages, then we add all pages to the array

        if (pages <= SIMPLE_PAGINATION_ITEMS) {
            range(1, pages + 1).map((pageNumber: number) => {
                addPage(pageNumber);
            });
        }

        // If the number of pages exceeds the maximum number of displayed pages

        if (pages > SIMPLE_PAGINATION_ITEMS) {
            // If the current page is less than the maximum number of pages displayed at the beginning before the spacer,
            // we show the couple pages, spacer and the last page
            if (currentPage < 5) {
                range(1, 6).map((pageNumber: number) => {
                    addPage(pageNumber);
                });
                addSpacer();
                addPage(pages);
            }

            // If the current page is greater than the maximum number of pages that are shown at the beginning
            // and less than the maximum number of pages that are shown at the end
            // to show the first page, a spacer, three pages of the current, spacer, and the last page
            if (currentPage > pages - 4) {
                addPage(1);
                addSpacer();
                range(pages - 4, pages + 1).map((pageNumber: number) => {
                    addPage(pageNumber);
                });
            }

            // If the current page is greater than the maximum number of pages displayed at the end after the spacer,
            // we show the first page, spacer and last pages
            if (currentPage > 4 && currentPage < pages - 3) {
                addPage(1);
                addSpacer();
                range(currentPage - 1, currentPage + 2).map((pageNumber: number) => {
                    addPage(pageNumber);
                });
                addSpacer();
                addPage(pages);
            }
        }

        return paginatorItems;
    }

    goToNext = () => {
        this.props.onValueChange(this.props.value + 1);
    }

    goToPrev = () => {
        this.props.onValueChange(this.props.value - 1);
    }

    render() {
        return this.props.render({
            size: this.props.size,
            pages: this.getPagesView(),
            goToNext: this.goToNext,
            goToPrev: this.goToPrev,
            isFirst: this.isFirst(),
            isLast: this.isLast(),
        });
    }
}