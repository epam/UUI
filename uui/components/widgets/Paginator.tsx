import React from 'react';
import cx from 'classnames';
import type { Overwrite } from '@epam/uui-core';
import { Paginator as UuiPaginator, PaginatorRenderParams, PaginatorProps as UuiPaginatorProps } from '@epam/uui-components';
import { Button } from '../buttons/Button';
import { settings } from '../../settings';

import css from './Paginator.module.scss';

export interface PaginatorModsOverride {}

interface PaginatorMods {
    /**
     * Defines component size
     *  @default '30'
     * */
    size?: '24' | '30';
}

export interface PaginatorProps extends UuiPaginatorProps, Overwrite<PaginatorMods, PaginatorModsOverride> {}

export function Paginator(props: PaginatorProps) {
    const renderPaginator = (params: PaginatorRenderParams) => (
        <nav role="navigation" className={ cx('uui-paginator', `uui-size-${props.size || settings.paginator.sizes.default}`) } { ...params.rawProps }>
            <ul
                className={ css.root }
            >
                <li>
                    <Button
                        size={ props.size || settings.paginator.sizes.default }
                        icon={ settings.paginator.icons.leftArrowIcon }
                        onClick={ params.goToPrev }
                        isDisabled={ params.isFirst || props.isDisabled }
                        fill="outline"
                        color="secondary"
                        rawProps={ {
                            'aria-label': 'Previous page',
                        } }
                    />
                </li>
                {params.pages.map((page, index) => {
                    if (page.type === 'spacer') {
                        return (
                            <li
                                key={ `${index}_spacer` }
                            >
                                <Button
                                    cx={ cx(css.spacer, css.page) }
                                    size={ props.size || settings.paginator.sizes.default }
                                    caption="..."
                                    fill="ghost"
                                    color="secondary"
                                    tabIndex={ -1 }
                                    isDisabled={ props.isDisabled }
                                />
                            </li>
                        );
                    } else {
                        return (
                            <li
                                key={ page.pageNumber }
                            >
                                <Button
                                    cx={ cx(css[`mode-${!page.isActive && 'ghost'}`], css.page) }
                                    size={ props.size || settings.paginator.sizes.default }
                                    caption={ page.pageNumber }
                                    onClick={ () => page.onClick?.() }
                                    rawProps={ { 'aria-current': page.isActive } }
                                    fill={ (page.isActive && 'outline') || 'ghost' }
                                    color="primary"
                                    isDisabled={ props.isDisabled }
                                />
                            </li>
                        );
                    }
                })}
                <li>
                    <Button
                        size={ props.size || settings.paginator.sizes.default }
                        icon={ settings.paginator.icons.rightArrowIcon }
                        onClick={ params.goToNext }
                        isDisabled={ params.isLast || props.isDisabled }
                        fill="outline"
                        color="secondary"
                        rawProps={ {
                            'aria-label': 'Next page',
                        } }
                    />
                </li>
            </ul>
        </nav>
    );

    return <UuiPaginator { ...props } render={ renderPaginator } />;
}
