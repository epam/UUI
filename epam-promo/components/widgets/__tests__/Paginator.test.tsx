import React from 'react';
import { Paginator } from '../Paginator';
import renderer from 'react-test-renderer';

describe('Paginator', () => {
    it('should be rendered correctly', () => {
        const tree = renderer.create(<Paginator value={null} onValueChange={() => {}} totalPages={10} size="24" />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
