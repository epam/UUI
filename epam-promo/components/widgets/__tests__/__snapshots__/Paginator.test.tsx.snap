// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator should be rendered correctly 1`] = `
<nav
  className="root"
  role="navigation"
>
  <button
    aria-disabled={false}
    className="container uui-button-box uui-enabled -clickable button-color-gray50 root size-24 fill-white fill-white size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <div
      className="container uui-icon uui-enabled"
      style={Object {}}
    >
      <svg
        className=""
      />
    </div>
  </button>
  <button
    className="container uui-button-box uui-enabled -clickable button-color-blue root size-24 fill-light fill-light size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <div
      className="uui-caption"
    >
      1
    </div>
  </button>
  <button
    className="container uui-button-box uui-enabled -clickable button-color-blue root size-24 fill-light fill-light size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <div
      className="uui-caption"
    >
      2
    </div>
  </button>
  <button
    className="container uui-button-box uui-enabled -clickable button-color-blue root size-24 fill-light fill-light size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <div
      className="uui-caption"
    >
      3
    </div>
  </button>
  <button
    className="container uui-button-box uui-enabled -clickable button-color-blue root size-24 fill-light fill-light size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <div
      className="uui-caption"
    >
      4
    </div>
  </button>
  <button
    className="container uui-button-box uui-enabled -clickable button-color-blue root size-24 fill-light fill-light size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <div
      className="uui-caption"
    >
      5
    </div>
  </button>
  <button
    className="container uui-button-box uui-enabled button-color-blue root size-24 fill-light fill-light size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={-1}
    type="button"
  >
    <div
      className="uui-caption"
    >
      ...
    </div>
  </button>
  <button
    className="container uui-button-box uui-enabled -clickable button-color-blue root size-24 fill-light fill-light size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <div
      className="uui-caption"
    >
      10
    </div>
  </button>
  <button
    aria-disabled={false}
    className="container uui-button-box uui-enabled -clickable button-color-gray50 root size-24 fill-white fill-white size-24"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    <div
      className="container uui-icon uui-enabled"
      style={Object {}}
    >
      <svg
        className=""
      />
    </div>
  </button>
</nav>
`;
